modules/nextflow/src/main/groovy/nextflow/extension/PublishOp.groovy

```diff
@@ -149,38 +202,46 @@ class PublishOp {
      * work directory paths to publish paths.
      *
      * @param value
+     * @param targetDirOrClosure
      */
-    protected Object normalizePaths(value) {
+    protected Object normalizePaths(value, targetDirOrClosure) {
         if( value instanceof Path ) {
-            return List.of(value.getBaseName(), normalizePath(value))
+            return List.of(value.getBaseName(), normalizePath(value, targetDirOrClosure))
         }
 
         if( value instanceof Collection ) {
             return value.collect { el ->
                 if( el instanceof Path )
-                    return normalizePath(el)
+                    return normalizePath(el, targetDirOrClosure)
                 if( el instanceof Collection<Path> )
-                    return normalizePaths(el)
+                    return normalizePaths(el, targetDirOrClosure)
                 return el
             }
         }
 
         if( value instanceof Map ) {
-            return value.collectEntries { k, v ->
-                if( v instanceof Path )
-                    return List.of(k, normalizePath(v))
-                if( v instanceof Collection<Path> )
-                    return List.of(k, normalizePaths(v))
-                return List.of(k, v)
-            }
+            return value
+                .findAll { k, v -> v != null }
+                .collectEntries { k, v ->
+                    if( v instanceof Path )
+                        return List.of(k, normalizePath(v, targetDirOrClosure))
+                    if( v instanceof Collection<Path> )
+                        return List.of(k, normalizePaths(v, targetDirOrClosure))
+                    return List.of(k, v)
+                }
         }
 
         throw new IllegalArgumentException("Index file record must be a list, map, or file: ${value} [${value.class.simpleName}]")
     }
 
-    private Path normalizePath(Path path) {
+    private String normalizePath(Path path, targetDirOrClosure) {
+        if( targetDirOrClosure instanceof Closure )
+            return FileHelper.asPath(targetDirOrClosure.call(path.getName()).toString()).normalize().toUriString()
         final sourceDir = getTaskDir(path)
-        return targetDir.resolve(sourceDir.relativize(path)).normalize()
+        if( sourceDir == null )
+            return path.toUriString()
+        final targetDir = FileHelper.asPath(targetDirOrClosure.toString())
+        return targetDir.resolve(sourceDir.relativize(path)).normalize().toUriString()
     }
```

Comment:
This become very "dense". Adding a javadoc and some comments explaining the logic of different branches would help preventing future headaches

Reply:
Also wonder why returning a string instead of `Path` the underlying operation will not require converting again to a `Path`?

Reply:
Somehow I managed to convert a path to string and back to path twice in this function 😅 

I think I changed to return string to ensure that paths were serialized correctly. Taking a second look, I see that the json writer already handles this and the csv writer just needed a bit of extra logic to do the same.

Between those simplifications and some extra comments, hopefully the code is much easier to understand now